glean-core/csharp/Glean/Dispatchers.cs

```diff
@@ -130,140 +114,94 @@ internal static Task LaunchAPI(Action action)
+                // Set the flag here after the initial tasks are added,
+                // combined with the lock on taskQueue to ensure initial tasks
+                // are executed first and subsequent tasks get executed as
+                // tasks after that and not queued.
+                QueueInitialTasks = false;
 
-        /// <summary>
-        /// Cancels any pending background tasks.
-        /// </summary>
-        internal static void CancelBackgroundTasks()
-        {
-            // Calling Complete on the taskSchedulerPair prevents it from
-            // accepting any new tasks.
-            taskSchedulerPair.Complete();
+                preInitActionQueue.Clear();
+            }
 
-            // Cancel tasks associated with the tokenSource
-            tokenSource.Cancel();
+            // Await all of the tasks up to QueueProcessingTimeout
+            Task.WaitAll(tasks.ToArray(), QueueProcessingTimeout);
 
-            // Clear the taskQueue
-            taskQueue = new ConcurrentQueue<Action>();
+            // This must happen after `QueueInitialTasks = false` is run, or
+            // it would be added to a full task queue and be silently dropped.
+            if (overflowCount > 0)
+            {
+                //TODO GleanError.preinitTasksOverflow
+                //         .addSync(MaxQueueSize + OverflowCount)
+            }
         }
 
-        /// <summary>
-        /// Stops queueing tasks and processes any tasks in the queue. Since
-        /// QueueInitialTasks is set to false prior to processing the queue, and
-        /// this function launches tasks from the ExclusiveScheduler back onto
-        /// the ExclusiveScheduler, the tasks should execute in order before any
-        /// new tasks are executed.
-        /// </summary>
-        internal static void FlushQueuedInitialTasks()
+        internal static Task ExecuteTask(Action action)
         {
-            // Set the flag first to guarantee new tasks are executed after
-            // this one.
-            QueueInitialTasks = false;
 
-            var task = Task.Factory.StartNew(() =>
+            if (TestingMode)
             {
-                // Add the tasks to the ExclusiveScheduler and execute them
-                // synchronously
-                while (taskQueue.TryDequeue(out Action localValue))
-                {
-                    localValue.Invoke();
-                }
-            });
-
-            // Wait for the initial tasks to execute
-            task.Wait(QueueProcessingTimeout);
-
-            if (OverflowCount > 0)
+                // If we are in testing mode, then await the task and return
+                // null
+                factory.StartNew(action).Wait();
+                return null;
+            }
+            else
             {
-                //TODO GleanError.preinitTasksOverflow
-                //         .addSync(MaxQueueSize + OverflowCount)
+                return factory.StartNew(action);
```

Comment:
Does this guarantee that `action` will be executed *right after* the current dispatched action, instead of being put at the end of the queue?

If not, this might require slightly changing `LimitedConcurrencyLevelTaskScheduler`.

Reply:
It appears to work this way from the only context in which it is called, which is inside of initialize, before queueing is turned off, so it should always be called when there are no other tasks queued.

Reply:
I removed any other call to it and replaced them with just invoking the action directly.

Reply:
I went ahead an changed this to not use the custom factory and instead use the default one and just throw this on the ThreadPool to ensure it gets executed off of the calling thread, and not on the queue.

Reply:
Actually, the more I think about this...  this should be launched on the apiScheduler but it's good to call out that Initialize is the only thing using this just to sidestep the pre-init queuing.

Reply:
Because that way it guarantees that initialize (and FlushQueuedTasks) get's called before anything else gets into the apiScheduler.

Reply:
Please note that this **has** to get executed _before_ any task in queue (except the one that's currently being executed). Is this checked for by any of our tests already? Did you check this manually?

> Actually, the more I think about this... this should be launched on the apiScheduler but it's good to call out that Initialize is the only thing using this just to sidestep the pre-init queuing.

This will change once we'll add the metrics ping scheduler. That's why we need to get this right. I'm fine with not keeping this PR waiting, we should file a bug and deal with this specific aspect later.